### DISCUSSION TITLE
EGraph Interactive Visualizer Research

### DISCUSSION THREAD

**Discussion by saulshanabrook** [saulshanabrook/saulshanabrook#34] [Blog]
I am starting to work on an interactive e-graph visualizer because I believe I am hitting some constraints with the current graphviz rendering. 

For example, I am trying to debug why the computation for [this e-graph](https://github.com/egraphs-good/egglog-python/blob/main/python/egglog/exp/array_api_loopnest.py) didn't fully complete as I imagined:

```python
>>> egraph.extract(val.shape[0])
TupleInt.range(Int(4)).filter(lambda i: ~TupleInt.from_vec(Vec[Int](Int(0), Int(1))).contains(i)).map(lambda i: TupleInt.from_vec(Vec[Int](Int(3), Int(2), Int(3), Int(4)))[i])[
    Int(0)
]
```

Like where did the rewrites fall down so that this isn't simplified to a number?

![output](https://github.com/user-attachments/assets/29d7445d-bfb0-489d-8f4c-2e5ebd99ae14)

So I thought I would go and see if I could build an interactive visualizer, instead of using graphviz. I had built [something similar for the Vega editor](https://github.com/vega/editor/pull/1023). So I know it's generally possible to build an interactive visualizer for large program graphs. I started looking at [all the code I wrote for that feature](https://github.com/vega/editor/tree/master/src/features/dataflow), and started to remember why it took so long... The combination of wrangling Cytoscape JS and Eclipse Layout Kernel ended up being quite a bit more work than one might imagine... Syncing all the state, doing things like pre-computing the expected size of each node based on the text layout so that it could assist in layout etc.

So I wanted to see, maybe there is an existing system that would work for me?

My requirements:

* Uses [Eclipse Layout Kernel](https://eclipse.dev/elk/)'s [layered layout](https://eclipse.dev/elk/reference/algorithms/org-eclipse-elk-layered.html). I have found this to work well for large numbers of nodes... It supports compound nodes and and just does a better layout than anything else I could find
* Supports compound nodes, so we can have e-classes and e-nodes inside of them.
* Supports "ports" meaning so that we have edges come from different parts of the node, corresponding to the different args to a function. ELK will take this into account to minimize crossing.
* Vaguely performant enough so we can visualize graphs with at least 1000k nodes.

Starting to look at [all the packages that depend on elkjs](https://github.com/kieler/elkjs/network/dependents?dependent_type=PACKAGE), I found a few that have some stars:

* [Mermaid](https://mermaid.js.org/) - no supports for ports https://github.com/mermaid-js/mermaid/issues/3169
* [Wing](https://github.com/winglang/wing) - used for viz for bespoke programming editor - not general purpose
* [unovis](https://github.com/f5/unovis) - "is a modular data visualization framework for React, Angular, Svelte, Vue, and vanilla TypeScript or JavaScript" - no support for ports
* [galaxy](https://github.com/galaxyproject/galaxy) some kind of science infrastructure - not general purpose
* [Reaflow](https://github.com/reaviz/reaflow) - Seems to [work with elk and custom ports](https://reaflow.dev/?path=/story/demos-layouts--custom-options)
* [Sprotty](https://github.com/eclipse-sprotty/sprotty) - official ELKjs visualizer... Could work? I remember last time it was a little confusing how to use this without a Java backend
* [Cytoscape JS ELK](https://github.com/cytoscape/cytoscape.js-elk) - official cytoscape ELKjs layout engine. I remember there were a few things missing here last time I used it so ended up re-implementing it.
* React Flow - Has no built in ELK integration but [comes with a demo using it for layout](https://reactflow.dev/learn/layouting/layouting#elkjs)

So I think for now, I will tryout reaflow and see if it doesnt't work with large graphs?

**Comment by shapr** [comment]
Have you already seen https://egraphs-good.github.io/egraph-visualizer/ ?
At some point I tracked down the julia library used to make the SVGs at the top of https://egraphs-good.github.io/ but I can't find it now.

**Reply by saulshanabrook** [reply]
Ha, thanks, yeah that's what I made in response to this post :)

If you output to the e-graph serialized format from egg you should be able to use the graphviz implementation I have there to get SVGs: https://github.com/egraphs-good/egraph-serialize/blob/main/src/graphviz.rs

I should add better docs or an easier way to do this from egg...
